images/hub/Dockerfile

```diff
@@ -45,11 +45,6 @@ RUN PYCURL_SSL_LIBRARY=openssl pip3 install --no-cache-dir \
             echo jupyterhub==${JUPYTERHUB_VERSION}; \
           fi')
 
-ADD jupyterhub_config.py /srv/jupyterhub_config.py
-
-ADD z2jh.py /usr/local/lib/python3.6/dist-packages/z2jh.py
-ADD cull_idle_servers.py /usr/local/bin/cull_idle_servers.py
```

Comment:
Do these other two files need to be dynamically mounted instead of being embedded in the image?

Reply:
Thanks @manics for helping out with reviewing this!

None of these three files need to be mounted as compared to be part of the image, but at least z2jh.py should go alongside like jupyterhub_config.py does as it is only used by jupyterhub_config.py and are greatly related.

But, I think all should be mounted instead of part of the image. It allows for:
1. Less need to rebuild this image in our CI job builds, it only needs to happen when we want to update the dependencies we install. This also increase local development speed significantly.
2. Whoever that use a custom image can keep in sync a lot easier as this person only needs to keep track of version changes as compared to also keeping track of any Helm chart configuration change that leads to a change in jupyterhub_config.py or similar.
3. This can help demonstrate a good practice on how to add a custom service in a sustainable way, where you would mount the additional source code files onto the container instead of rebuilding the hub image with the added service within it.

I think all of this leads to a better Developer Experience and less potential issues for helm chart users.

Reply:
That's arguably the case for `z2jh.py` though it feels wrong to be dynamically mounting a Python module into `{dist,site}-packages`.

In the case of `cull_idle_servers.py` isn't that tied to the JupyterHub API, so if for example the script was updated to make use of a new JupyterHub API feature then someone running an older hub image with a newer chart could end up with an incompatible `cull_idle_servers.py`?

Reply:
Ah, yes I agree about mounting the module into `{dist,site}-packages`, it was what was done in the Docker image so I figured I'll simply follow that practice. I'm not yet python savvy enough to come up with a suggestion about where it should be located to be accessible from jupyterhub_config.py in a practical enough way. Got ideas?

Reply:
Regarding `cull_idle_servers.py`, it has ties to the JupyterHub API:

```python
    c.JupyterHub.services.append({
        'name': 'cull-idle',
        'admin': True,
        'command': cull_cmd,
    })
```

Most of its API ties relates to the Helm chart I'd say, as the scripts is configured with various parameters passed to it through the Helm chart etc.

Reply:
I had a think about this, and I think we're missing instructions (or a set or requirements) for someone wanting to use a custom hub image.

Currently a custom hub image should have:
- a compatible version of jupyterhub
- compatible versions of any plugins such as Authenticators that are used
- the `cull_idle_servers.py` script
- the `z2jh.py` module or some other way of parsing the configuration properties
- Anything else? Does z2jh make any other assumptions?

In practice this is best done by extending `FROM` the official chart image

With this PR a custom hub image should have:
- a compatible version of jupyterhub
- compatible versions of any plugins such as Authenticators that are used
- no `cull_idle_servers.py` script as it'll be overwritten
- no `z2jh.py` module as it'll be overwritten
- Anything else? Does z2jh make any other assumptions?

This means you can have a custom hub image built from scratch.

Both these are reasonable (ignore for now the question of where the files are mounted), but if someone has a custom `cull_idle_servers.py` script this PR is arguably a breaking change (add to the release notes?).

Reply:
Thanks @manics for your review and time spent thinking on this! I created https://github.com/jupyterhub/zero-to-jupyterhub-k8s/issues/1424 to represent this documentation!